### ISSUE TITLE
Header - Coveo search

### ISSUE THREAD

**Issue by mhaack** [hlxsites/danaher-ls-aem#27]
The main search bar in the header is implemented via [Coevo Atomic library](https://docs.coveo.com/en/atomic/latest/) which we need to implement as well. All search requests will got to Coevo. 

This is blocked by #6 which should be implemented first.

(Currently we don't index Franklin pages to Coevo, this will happen on a later stage)

**Comment by mhaack** [comment]
Did some quick test
```<html>
  <head>
    <title>MH Test</title>
    <script>
          (async () => {
            await customElements.whenDefined('atomic-search-interface');
            const searchInterface = document.querySelector('#search');

            await searchInterface.initialize({
              accessToken:'...',
              organizationId: '...',
              organizationEndpoints: await searchInterface.getOrganizationEndpoints('...'),
            });
            searchInterface.executeFirstSearch();
          })();
    </script>
  </head>
  <body>
    <div style="height: 100px;">
        <script type="module" src="https://static.cloud.coveo.com/atomic/v2/atomic.esm.js"></script>
        <style rel="stylesheet" href="https://static.cloud.coveo.com/atomic/v2/themes/coveo.css"></style>
        <atomic-search-interface
            class="search-bar hydrated"
            search-hub="DanaherMainSearch"
            pipeline="Danaher Marketplace"
            language="en"
            reflect-state-in-url=""
            scroll-container="atomic-search-interface"
            language-assets-path="./lang"
            icon-assets-path="./assets"
            enable-relevance-inspector=""
            id="search">
            <atomic-search-box
            redirection-url="https://lifesciences.danaher.com/us/en/search.html"
            data-atomic-rendered="true"
            data-atomic-loaded="true"
            class="hydrated"
            number-of-queries="8"
            minimum-query-length="0"
            clear-filters="">
                <atomic-search-box-recent-queries max-with-query="3" class="hydrated"></atomic-search-box-recent-queries>
                <atomic-search-box-query-suggestions class="hydrated"></atomic-search-box-query-suggestions>
            </atomic-search-box>
        </atomic-search-interface>
    </div>
  </body>
</html>
```
While this is super easy to integrate it loads 237kBytes compressed. A lot for just the search bar. We should consider using the APIs. @davenichols-DHLS @rgravitvl any thoughs?

**Comment by davidzochowski** [comment]
Here is the information about how we can implement Coveo.

1. Atomic
Information:
https://docs.coveo.com/en/atomic/latest/
https://docs.coveo.com/en/atomic/latest/usage/
Online Course:
https://levelup.coveo.com/learn/courses/atomic

This is Coveo's Component Library that contains pre-built UI component that can be placed onto a page.
This contains html, css, and js.
It is built on Coveo's Headless Library so when using Atomic, are loading both Atomic and Headless libraries.
As this is built on top of Headless, it has analytics built-in so there is no need to develop analytics unless something custom is needed.
The main benefit of this is the short implementation time as a bulk of the work is already done and all we need to do is slight tweaks and some styling changes.
The main detriment of this is that since we are relying on Coveo's pre-built components, customization of these components would take some time as the component would need to be overwritten. Also, this requires the most kb to download to a user's browser so this could impact pagespeed.


2. Headless
Information:
https://docs.coveo.com/en/headless/latest/
https://docs.coveo.com/en/headless/latest/usage/
Online Course:
https://levelup.coveo.com/learn/courses/headless-commerce-tutorial

This is Coveo's JavaScript Library specifically to query their API and fetch the results.
There is analytics built in so there should be no need to develop anything alaytics related unless something custom is needed.
The main benefit of this is that since this is only Javascript, we would retain full control and customization of the UI components.
The main detriment is that there would be a longer development period as the UI would need to be built.


3. Direct API Calls
This would be directly calling Coveo's API to query and return results.
Since this has no libraries, we would need to build out the analytics code. We can use the coveoua library mentioned below but we would still need to incorporate it.
The main benefit of this is that there are no extra libraries required so this would give the biggest benefit to pagespeed.
The main detriment of this is that, as this is completely custom, there would be an even longer development time.


Extra:

1. coveoua
This is Coveo's Javascript Library specifically for analytics.
https://docs.coveo.com/en/l2bf0354/coveo-for-commerce/initialize-the-coveo-analytics-library
We would need to use this because while Headless can take care of analytics whenever the user interacts with the search aspects of the site, we want to send information to Coveo for pages that have no search interaction.
Examples: 

- Page Views
- Product Views
- Ecommerce events (add to cart, remove from cart, checkout, purchase)

Since this does not change the functionality of a page from a user perspective, this library can be loaded in the footer so the DOM does not have to wait for it to finish.

**Comment by duynguyen** [comment]
Acceptance checklist:
[] click on the searchbox should show default suggestions
[] empty search term should show 3 recent searches on top
[] clear icon should appear when search term is not empty
[] submit with empty search (enter or click search button) should go to search page without query
[] submit with a search term (enter or click search button) should go to search page with a query
[] click on a search suggestion should go to search page with the selected query
[] search suggestions can be navigated using up / down keys
[] press enter key on a search suggestion should go to search page with the selected query
[] search suggestion should include analytics of the current user session
[] search submission should record the search call
[] upon a search submission with a query, the search term is recorded in recent searches list
[] click outside the search box & suggestions should close the suggestions list
